src/order/height.lean

```diff
@@ -0,0 +1,276 @@
+/-
+Copyright (c) 2022 Andrew Yang. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Andrew Yang
+-/
+
+import data.list.chain
+import data.nat.lattice
+import algebra.order.monoid
+
+/-!
+
+# Maximal length of chains
+
+This file contains lemmas to work with the maximal length of strictly ascending finite
+sequences (chains) in a partially ordered set.
+
+## Main definition
+
+- `sup_chain_length α`: The maximal length of chains in `α`.
+- `chain_height a`: The maximal length of chains in `α` that start from `a : α`
+  (not including `a` itself).
+
+They are both valued in `enat`, and equal `⊤ : enat` when the lengths are unbounded.
+
+## Main results
+
+- `exists_chain_of_le_length`: For each `n : ℕ` such that `n ≤ chain_height a`, there exists a
+  chain of length `n` starting from `a`.
+- `chain_height_eq_one_iff`: An element has height zero iff it is minimal.
+- `chain_height_le_of_lt`: If `a < b`, then `chain_height b + 1 ≤ chain_height a`. This implies that
+  `chain_height` is decreasing (`chain_height_antitone`).
+- `sup_chain_length_eq_sup_chain_height`: `sup_chain_length` is the supremum of all the heights.
+- `sup_chain_length_dual`: The maximal length of a poset and is dual is the same.
+- `chain_height_add_chain_height_dual_le`: The sum of the height of the same element in `α` and
+  `αᵒᵈ` is bounded by `sup_chain_length`.
+
+-/
+
+universes u v
+
+namespace list
+
+variables (α : Type u) [partial_order α]
+
+/-- The maximal length of a strictly ascending sequence in a poset. -/
+noncomputable
+def sup_chain_length : enat := supr $ λ l : { l : list α // list.chain' (<) l }, l.1.length
+
+variable {α}
+
+/-- The maximal length of a strictly ascending sequence in a poset starting from `a`.
+Note that `a` itself is not counted in the length. -/
+noncomputable
+def chain_height (a : α) : enat := supr $ λ l : { l // list.chain (<) a l }, l.1.length
```

Comment:
I would expect this to be the other way around, so that `⊥` has height zero and `⊤` has height `sup_chain_length α`. This would better match what we have going on `grade_order`.